Cython/Build/Cache.py

```diff
@@ -0,0 +1,189 @@
+
+    """
+    if 'CYTHON_CACHE_DIR' in os.environ:
+        return os.environ['CYTHON_CACHE_DIR']
+
+    parent = None
+    if os.name == 'posix':
+        if sys.platform == 'darwin':
+            parent = os.path.expanduser('~/Library/Caches')
+        else:
+            # this could fallback on ~/.cache
+            parent = os.environ.get('XDG_CACHE_HOME')
+
+    if parent and os.path.isdir(parent):
+        return os.path.join(parent, 'cython')
+
+    # last fallback: ~/.cython
+    return os.path.expanduser(os.path.join('~', '.cython'))
+
+
+class Cache:
+
+    def __init__(self, cache, cache_size=None):
+        if cache is True:
+            self.enabled = True
+            self.path = os.path.join(get_cython_cache_dir(), 'compiler')
+        elif cache:
+            self.path = cache
+            self.enabled = True
+        else:
+            self.path = None
+            self.enabled = False
+        self.cache_size = cache_size if cache_size is not None else 1024 * 1024 * 100
+        if self.enabled:
+            if not os.path.exists(self.path):
+                os.makedirs(self.path)
+
+    def transitive_fingerprint(self, filename, dependencies, compilation_options, language, py_limited_api=False, np_pythran=False):
+        r"""
+        Return a fingerprint of a cython file that is about to be cythonized.
+
+        Fingerprints are looked up in future compilations. If the fingerprint
+        is found, the cythonization can be skipped. The fingerprint must
+        incorporate everything that has an influence on the generated code.
+        """
+        try:
+            m = hashlib.sha1(__version__.encode('UTF-8'))
+            m.update(file_hash(filename).encode('UTF-8'))
+            for x in sorted(dependencies.all_dependencies(filename)):
+                if os.path.splitext(x)[1] not in ('.c', '.cpp', '.h'):
+                    m.update(file_hash(x).encode('UTF-8'))
+            # Include the module attributes that change the compilation result
+            # in the fingerprint. We do not iterate over module.__dict__ and
+            # include almost everything here as users might extend Extension
+            # with arbitrary (random) attributes that would lead to cache
+            # misses.
+            m.update(
+                str((language, py_limited_api,np_pythran)).encode('UTF-8')
+            )
+            m.update(compilation_options.get_fingerprint().encode('UTF-8'))
+            return m.hexdigest()
+        except OSError:
+            return None
+
+    def fingerprint_file_base(self, cfile, fingerprint):
+        return join_path(self.path, "%s-%s" % (os.path.basename(cfile), fingerprint))
+
+    def gz_fingerprint_file(self, cfile, fingerprint):
+        return self.fingerprint_file_base(cfile, fingerprint) + gzip_ext
+
+
+    def zip_fingerprint_file(self, cfile, fingerprint):
+        return self.fingerprint_file_base(cfile, fingerprint) + '.zip'
+
+    def lookup_cache(self, c_file, fingerprint):
```

Comment:
This method doesn't just do a lookup, it extracts files from the cache. Funnily, it also creates directories, even if it doesn't find anything in the cache. And it then even returns True/False to indicate whether it did something or not. Loads of functionality stuffed into one function.

I'd split this into two methods, one that does the lookup and returns the cache file or None, and one that extracts the files from the cache file. That way, users are informed after the first call that a cache entry exists, and can call the second method to use it.

Reply:
Fixed in https://github.com/cython/cython/pull/6117.